Support Gorouter talking to NATS over mTLS

46bit opened the pull request:
* *A short explanation of the proposed change:* this PR adds boshrelease support for mTLS between Gorouter and NATS. This PR sits alongside https://github.com/cloudfoundry/gorouter/pull/283.

* *An explanation of the use cases your change solves:* 

* *Instructions to functionally test the behavior change using operator interfaces (BOSH manifest, logs, curl, and metrics):* a cf-deployment PR is coming with which this can be deployed and observed.

* *Links to any other associated PRs*: https://github.com/cloudfoundry/gorouter/pull/283 https://github.com/cloudfoundry/cf-deployment/issues/906

* [x] *I have viewed signed and have submitted the Contributor License Agreement:* we are actively working on this.

* [x] *I have made this pull request to the `develop` branch*

* [x] *I have run all the unit tests using `scripts/run-unit-tests-in-docker`*

* [ ] (Optional) I have run Routing Acceptance Tests and Routing Smoke Tests on bosh lite

* [ ] (Optional) I have run CF Acceptance Tests on bosh lite

Commit:
Configure Gorouter to talk to NATS over mTLS

Until now, Gorouter has talked to NATS over an unencrypted connection.
Authentication was with a password in plaintext. We want to encrypt
and authenticate this connection. For that purpose we're using Mutual
TLS, the same as route-emitter and route-registrar use with NATS.

This commit passes Gorouter the properties it needs from BOSH. The YAML
used for Gorouter's NATS Configuration changes quite a bit as part of this,
so that we don't have to include N copies of the TLS certificates. But the
only new limitation introduced is that all NATS servers have to use the same
username and password.

Commit:
Be able to use hostname TLS certs for NATS

The TLS certificate presented by NATS is issued for its DNS name, and
not for the individual NATS server IP addresses. Gorouter was being
configured to connect to IP addresses, not the DNS name, and so the
TLS connection failed.

This code is borrowed from route-emitter where someone already had
to solve this [1].

[1] https://github.com/cloudfoundry/diego-release/commit/44260f2401c346dd2d5f73e9587fd88c2b86007e

ameowlia commented:
Hi @46bit, 

Thank you so much for your PRs!! We really appreciate it. 

Just glancing over the code, it all looks sane to me, but I'd like to request a few additions: 

* Can you please add template tests for the new properties? They live [here](https://github.com/cloudfoundry/routing-release/blob/develop/spec/gorouter_templates_spec.rb) and are run with rspec. They also run automatically when you run `scripts/run-unit-tests-in-docker`. 
* Can you add an opsfile that turns on this feature and generates a client cert? [Here](https://github.com/cloudfoundry/routing-release/blob/develop/ci/opsfiles/README.md) is where we keep the opsfiles. And [here](https://github.com/cloudfoundry/routing-release/blob/develop/ci/opsfiles/add-routing-api-mtls-certificates.yml) is an example opsfile that generates certificates as bosh variables.

✨ Thank you!

46bit commented:
You're welcome @ameowlia ✨ Very happy we're contributing this!

I'll get working on your suggestions. I'm also going to run things like the Routing Acceptance Tests to check they pass.

Commit:
Bring NATS TLS Opsfile up to date

Since this file was created, route_registrar and NATS are now using TLS in cf-deployment.
Therefore most of this Opsfile is now unnecessary. The nats-smoke-test is retained because
it does not appear to be part of cf-deployment.

This commit forms part of a PR in which we're making Gorouter and NATS talk via TLS, so
the Opsfile now configures that.

Commit:
Add NATS TLS manifests tests and refactor

46bit commented on spec/gorouter_templates_spec.rb:
```diff
@@ -777,6 +781,231 @@ def make_properties(link, link_property)
+            def make_properties(link, link_property)
+              property_chain = link.split('.').reverse
+
+              property_chain.reduce(link_property) do |memo, obj|
+                { obj => memo }
+              end
+            end
+
+            let(:links) do
+              [
+                Bosh::Template::Test::Link.new(
+                  name: link_config.link_namespace,
+                  properties: make_properties(link_config.link, link_value)
+                )
+              ]
+            end
+
+            let(:rendered_template) { template.render(deployment_manifest_fragment, consumes: links) }
+
+            context 'when the property is set' do
+              before do
+                set_at_property(deployment_manifest_fragment, link_config.property, property_value)
+              end
+
+              it 'should prefer the value in the properties' do
+                expect(get_at_property(parsed_yaml, link_config.parsed_yaml_property)).to eq(property_value)
+              end
+            end
+
+            context 'when the property is not set' do
+              before do
+                delete_at_property(deployment_manifest_fragment, link_config.property)
+              end
+
+              it 'should render the value from the link' do
+                expect(get_at_property(parsed_yaml, link_config.parsed_yaml_property)).to eq(link_value)
+              end
+            end
+          end
+        end
```
I anticipate a reviewer will wonder why I duplicated this from further down in the file. I've tried moving this code several times and kept hitting intractable problems with the scoping of `deployment_manifest_fragment` and how RSpec shared examples work.

ameowlia replied:
👍

46bit commented:
Hey @ameowlia. I've added commits with your suggested additions. Also a commit to fix a linting test failure from #202. Ready for re-review :)

46bit commented:
Latest changes broke something. I'll investigate on Monday.

ameowlia commented:
Thanks for your linting fix 🤦‍♀️ . I'll review the code when it's non-breaking. Let me know when that is.

46bit commented:
@ameowlia fixed it. Still waiting on CLA but it'd be great to get feedback while we're waiting

ameowlia commented on ci/opsfiles/enable-nats-tls-for-cf.yml:
```diff
@@ -1,158 +1,9 @@
-  path: /addons?/name=bosh-dns-aliases/jobs/name=bosh-dns-aliases/properties/aliases/domain=nats.service.cf.internal
+  path: /instance_groups/name=router/jobs/name=gorouter/properties/nats?
   value:
-    domain: nats.service.cf.internal
```
❓ why are you deleting so much here? Is it all default in cf-d now?

michaelmccaskill replied:
As of cf-deployment [v12.44.0](https://github.com/cloudfoundry/cf-deployment/releases/tag/v12.44.0), that is correct. That version was released over a year ago so it's probably ok to remove it from here.

ameowlia reviewed (commented):
Just one question about the opsfile, then I think it looks good to go.

WillsonHG commented:
/easycla

46bit commented:
@ameowlia EasyCLA is signed off, so this can get moving again 🎉

46bit commented:
I reran the unit tests and some of the template specs in `spec/gorouter_templates_spec.rb` are failing. They were written assuming that the first token in the property name is the same as the link name, but that isn't true for `nats-tls`. Fixing this has turned into a multi-hour horror; I can't do it.

Suggest we drop the extra tests and sort this out when we get rid of the non-TLS NATS, when things will simplify. Thoughts?

ameowlia commented:
Hi @46bit 👋 

I spent some time trying to get this to pass today. [Here is a messy, but successful effort](https://github.com/cloudfoundry/routing-release/commit/6e41329f1344d0cdfeed9bf98e6480aacd6c6688).

**What I learned** 
* Problem 1: The duplicate names for the method (property_or_link) the class (LinkConfiguration), and the table test (overridable_link) interfered with each other. I appended "nats" onto the nats ones and left the ones for routing api alone so I could focus on just the nats stuff.
* This should be dried up.
* Problem 2: the bosh links were being created incorrectly
```
# correct
[#<Bosh::Template::Test::Link:0x00007fcc7cae8058 @instances=[], @name="nats-tls", @properties={"nats"=>{"external"=>{"tls"=>{"ca"=>"ihwgrnqdtvjuomzfbypxkcesal"}}}}>]

# incorrect
[#<Bosh::Template::Test::Link:0x00007fefda0d68b8 @instances=[], @name="nats-tls", @properties={"nats-tls"=>{"nats"=>{"external"=>{"tls"=>{"ca"=>"depstrkiqbzcnovhyjaufwlxgm"}}}}}>]
```
* I found it easiest to debug this by putting a `p` [here](https://github.com/cloudfoundry/routing-release/commit/710875597d46ca777cd6d55be2d641cc85564e49), so that way I could see what link was being generated. 


**Next steps** 
* Can you review/integrate/dry these changes to make the tests pass?

Thanks, as always, for all your hard work!

Commit:
Fix Gorouter property tests

This commit deduplicates and fixes tests added as part of allowing
Gorouter to talk to NATS over mTLS.

Huge thanks to @ameowlia for figuring out how to fix the tests.
https://github.com/cloudfoundry/routing-release/pull/204#issuecomment-875068923

46bit commented:
@ameowlia thank you!!

I've added an extra commit that fixes the tests and does some deduplicating :)

ameowlia commented:
🎉 🐱  Thank you so much @46bit!
